src/main/scala/workflow/WorkflowUtils.scala

```diff
@@ -0,0 +1,124 @@
+              fitDeps.append(None)
+            }
+            case estimatorFitNode: EstimatorFitNode => {
+              instructionIdToNodeId.put(instruction, nodes.length)
+              // TODO: Get more reasonable label...
+              nodes.append(new DelegatingTransformerNode("Estimator fit"))
+              dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+              fitDeps.append(Some(instructionIdToNodeId(transformer)))
+            }
+            case _ => throw new RuntimeException("Transformer apply instruction must point at a Transformer")
+          }
+        }
+        case EstimatorFitNode(est, inputs) => {
+          val estimatorNode = instructions(est).asInstanceOf[EstimatorNode]
+          instructionIdToNodeId.put(instruction, nodes.length)
+          nodes.append(estimatorNode)
+          dataDeps.append(inputs.map(instructionIdToNodeId.apply))
+          fitDeps.append(None)
+        }
+      }
+    }
+
+    new ConcretePipeline(nodes.toSeq, dataDeps.toSeq, fitDeps.toSeq, nodes.length - 1)
+  }
+
+  def pipelineToInstructions[A, B](pipeline: Pipeline[A, B]): Seq[Instruction] = {
+    val nodes = pipeline.nodes
+    val dataDeps = pipeline.dataDeps
+    val fitDeps = pipeline.fitDeps
+    val sink = pipeline.sink
+
+    pipelineRecurse(sink, nodes, dataDeps, fitDeps, Map(Pipeline.SOURCE -> Pipeline.SOURCE), Seq())._2
+  }
+
+  def pipelineRecurse(
+    current: Int,
+    nodes: Seq[Node],
+    dataDeps: Seq[Seq[Int]],
+    fitDeps: Seq[Option[Int]],
+    nodeIdToInstructionId: Map[Int, Int],
+    instructions: Seq[Instruction]
+  ): (Map[Int, Int], Seq[Instruction]) = {
+    var curIdMap = nodeIdToInstructionId
+    var curInstructions = instructions
+    System.out.println(current)
+    for (dep <- fitDeps(current)) {
+      if (!curIdMap.contains(dep) && dep != Pipeline.SOURCE) {
+        val (newIdMap, newInstructions) = pipelineRecurse(dep, nodes, dataDeps, fitDeps, curIdMap, curInstructions)
+        curIdMap = newIdMap
+        curInstructions = newInstructions
+      }
+    }
+
+    for (dep <- dataDeps(current)) {
+      if (!curIdMap.contains(dep) && dep != Pipeline.SOURCE) {
+        val (newIdMap, newInstructions) = pipelineRecurse(dep, nodes, dataDeps, fitDeps, curIdMap, curInstructions)
+        curIdMap = newIdMap
+        curInstructions = newInstructions
+      }
+    }
+
+    nodes(current) match {
+      case source: SourceNode => {
+        curIdMap = curIdMap + (current -> curInstructions.length)
+        curInstructions = curInstructions :+ source
+        (curIdMap, curInstructions)
+      }
+
+      case transformer: TransformerNode => {
+        curInstructions = curInstructions :+ transformer
+        val inputs = dataDeps(current).map(curIdMap.apply)
+        curIdMap = curIdMap + (current -> curInstructions.length)
+        curInstructions = curInstructions :+ TransformerApplyNode(curInstructions.length - 1, inputs)
```

Comment:
These three blocks are *almost* identical - is there a way this can be tightened?